ADd more explanation about rate limit (#2114)
TykTechnologies/tyk-docs: tyk-docs/content/getting-started/key-concepts/rate-limiting.md

---
title: "Rate Limiting in Tyk"
date: 2021-02-04

> tags: ["Rate Limit", "Rate Limiting", "Rate Limit Algos", "DRL"]

Reviewer: ```suggestion
tags: ["Rate Limit", "Rate Limiting", "Rate Limit Algorithms", "DRL"]
```

> description: "Overview on Rate Limiting with the Tyk Gateway"

Reviewer: ```suggestion
description: "Overview of Rate Limiting with the Tyk Gateway"
```

Rate limits are calculated in Requests Per Second (RPS). For example, let’s say a developer only wants to allow a client to call the API a maximum of 10 times per minute. In this case the developer would apply a rate limit to their API expressed as "10 requests per 60 seconds". This means that the client will be able to successfully call the API up to 10 times within any 60 second interval and after that the user will get an error stating their rate limit has been exceeded if they call it an 11th time within that time frame.

## Types Of Rate Limiting

Tyk offers 2 different rate limiting algorithms that the gateway uses and they behave in different ways.

> 1. Distributed Rate Limiter.  Most performant, not 100% perfect accuracy.  Recommended for most use cases.

Reviewer: ```suggestion
1. Distributed Rate Limiter.  Most performant, not 100% perfect accuracy.  Recommended for most use cases. Uses the leaky bucket algorithm.
```

> 1. Distributed Rate Limiter.  Most performant, not 100% perfect accuracy.  Recommended for most use cases. Uses the leaky bucket algorithm.

Reviewer: ```suggestion
1. Distributed Rate Limiter.  Most performant, not 100% perfect accuracy.  Recommended for most use cases. Uses the [leaky bucket algorithm](https://en.wikipedia.org/wiki/Leaky_bucket).
```

> 2. Redis Rate Limiter.  Less performant, 100% perfect accuracy.

Reviewer: ```suggestion
2. Redis Rate Limiter.  Less performant, 100% perfect accuracy. Uses the sliding window algorithm.
```

> 2. Redis Rate Limiter.  Less performant, 100% perfect accuracy. Uses the sliding window algorithm.

Reviewer: ```suggestion
2. Redis Rate Limiter.  Less performant, 100% perfect accuracy. Uses the [sliding window algorithm](https://developer.redis.com/develop/dotnet/aspnetcore/rate-limiting/sliding-window/).
```

#### Distributed Rate Limiter (DRL)

This is the default rate limiter in Tyk.  It is the most performant, and the trade-off is that the limit is approximate, not exact.  To use a less performant, exact rate limiter, review the Redis rate limiter below.

With the DRL, the gateways look at the rate per second and count up how many gateways there are (via shared redis) and divide the rate evenly between them. They keep the rate in memory and start sending `429`s when their share is used up.

> This relies on having a very fair load balancer since it assumes a well distributed load between all of the gateways.

Reviewer: ```suggestion
This relies on having a fair load balancer since it assumes a well distributed load between all of the gateways.
```

> It also uses what's called a leaky bucket algorythm. In this case if the request rate is higher than the rate limit it will attempt to let through requests at the specified rate limit. It's important to note that this is the only rate limit method that uses this algorythm and that it will yield approximate results.

Reviewer 2: add a link to this https://en.wikipedia.org/wiki/Leaky_bucket
isn't it `algorithm`

> It also uses what's called a leaky bucket algorythm. In this case if the request rate is higher than the rate limit it will attempt to let through requests at the specified rate limit. It's important to note that this is the only rate limit method that uses this algorythm and that it will yield approximate results.

Reviewer: ```suggestion
It also uses what's called a leaky bucket algorithm. In this case if the request rate is higher than the rate limit it will attempt to let through requests at the specified rate limit. It's important to note that this is the only rate limit method that uses this algorithm and that it will yield approximate results.
```

#### Redis rate limiter

> This uses redis to track the rate of incoming API calls. It's important to note that it blocks access to the API when the rate exceeds the rate limit. It's not like the leaky bucket algorythm, it doesn't let API calls through until the rate drops below the rate limit for the period of time that the limit uses. For example if the rate limit is 3000/minute the call rate would have to be reduced below 3000 for a whole minute before the 429s would stop.

Reviewer 2: not so clear `before the 429s would stop.` - clarify what is this  http error code  and add link https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429

> This uses redis to track the rate of incoming API calls. It's important to note that it blocks access to the API when the rate exceeds the rate limit. It's not like the leaky bucket algorythm, it doesn't let API calls through until the rate drops below the rate limit for the period of time that the limit uses. For example if the rate limit is 3000/minute the call rate would have to be reduced below 3000 for a whole minute before the 429s would stop.

Reviewer: ```suggestion
This uses redis to track the rate of incoming API calls. It's important to note that it blocks access to the API when the rate exceeds the rate limit. Unlike the leaky bucket algorithm, it doesn't let API calls through until the rate drops below the rate limit for the period of time that the limit uses. For example if the rate limit is 3000/minute the call rate would have to be reduced below 3000 for a whole minute before the 429s would stop.
```

#### DRL Threshold

> `TYK_GW_DRLTHRESHOLD`

Reviewer: We should also add the non-transactional to this.

Optionally, you can use both rate limit options simealtanoeusly.  This is suitable for hard-syncing rate limits for lower thresholds, ie for more expensive APIs, and using the more performant Rate Limiter for the higher traffic APIs.

Tyk switches between these two modes using the `drl_threshold`. If the rate limit is more than the drl_threshold (per gateway) then the DRL is used. If it's below the DRL threshold the redis rate limiter is used.

Read more [about DRL Threshold here]({{< ref "/content/tyk-stack/tyk-gateway/configuration/tyk-gateway-configuration-options.md#drl_threshold" >}})

> Redis rate limiter, which is 100% correct, but it also use dynamic window, which means that if there is user who abuse rate limit, it will not allow his requests, untill he start respecting rate limit. In other words failed rate limit attempts, also count towards rate limit.

Reviewer 2: isn't it a sliding window?
Also add link to this https://developer.redis.com/develop/dotnet/aspnetcore/rate-limiting/sliding-window/

> Redis rate limiter, which is 100% correct, but it also use dynamic window, which means that if there is user who abuse rate limit, it will not allow his requests, untill he start respecting rate limit. In other words failed rate limit attempts, also count towards rate limit.

Reviewer: ```suggestion
Redis rate limiter, is 100% correct, but it uses the sliding window algorithm instead of the leaky bucket. This means that if there is a user who abuses the rate limit, this user's requests will be limited until they start respecting the rate limit. In other words, requests that return 429 (Rate Limit Exceeded) will count towards their rate limit counter.
```

> Redis rate limiter, provides 100% accuracy, however instead of using the leaky bucket algorithm it uses the sliding window algorithm. This means that if there is a user who abuses the rate limit, this user's requests will be limited until they start respecting the rate limit. In other words, requests that return 429 (Rate Limit Exceeded) will count towards their rate limit counter.

Reviewer: ```suggestion
Redis rate limiter, provides 100% accuracy, however instead of using the leaky bucket algorithm it uses the sliding window algorithm. This means that if there is a user who abuses the rate limit, this user's requests will be limited until they start respecting the rate limit. In other words, requests that return [429 (Rate Limit Exceeded)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429) will count towards their rate limit counter.
```

> In case when you have small rate limit with big amount of servers, Gateway always switch to Redis rate limiter, which means that you can have only moving window algorithm in that case. And clients which abuse rate limit, need be aware about this behavior, or you need to increare rate limit for them, if you can't stop client from abusing rate limit.

Reviewer: I don't understand this.
